R/db.R

```diff
@@ -150,33 +191,146 @@ write_data_safely <- function(dbconn, tblname, df, key) {
     TRUE
   }
 
-  write_column_hashed_data <- function(df, append = TRUE) {
+  ## Input: table name and calculated shard names
+  write_table_shard_map <- function(tblname, shard_names) {
+    ## Example:
+    ##
+    ## |   | table_name  | shard_name  |
+    ## |---|-------------|-------------|
+    ## | 1 | tblname_1   | shard_1     |
+    ## | 2 | tblname_1   | shard_2     |
+    ## | 3 | tblname_2   | shard_3     |
+    table_shard_map <- data.frame(table_name = rep(tblname, length(shard_names)), shard_name = shard_names)
+    ## If we don't do this, we will get really weird bugs with numeric things stored as character
+    ## For example, a row with ID 100000 will be stored as 10e+5, which is wrong.
+    old_options <- options(scipen = 20, digits = 20)
+    on.exit(options(old_options))
+
+    ## Store the map of logical table names to physical shards in the table_shard_map table.
+    if (!DBI::dbExistsTable(dbconn, 'table_shard_map'))
+      dbWriteTableUntilSuccess(dbconn, 'table_shard_map', table_shard_map)
+    else {
+      shards <- DBI::dbGetQuery(dbconn, pp("SELECT shard_name FROM table_shard_map WHERE table_name='#{tblname}'"))
+      if (NROW(shards) > 0) {
```

Comment:
If `FALSE` for whatever reason, `table_shard_map` won't be in scope below, causing an error. Maybe 

```r
else {
  table_shard_map <- NULL
}
```